File: examples/go/01_basic_parse/main.go
package main

import (
	"fmt"

	"github.com/qdrant/qql/gqql"
)

func main() {
	ast, _ := gqql.Parse("CREATE COLLECTION docs HYBRID WITH HNSW (m = 32)")
	fmt.Println("=== Parsed AST ===")
	fmt.Println(trunc(ast, 500))

	fmt.Println("\n=== Tokens ===")
	tokens, _ := gqql.Tokenize("QUERY 'vector database' FROM docs LIMIT 10")
	fmt.Println(tokens[:truncIdx(tokens, 400)])

	fmt.Println("\n=== Validation ===")
	for _, q := range []string{
		"QUERY 'hello' FROM docs LIMIT 5",
		"CREATE COLLECTION docs",
		"SELECT * FROM docs WHERE id = 1",
		"",
		"BOGUS STUFF",
	} {
		fmt.Printf("  valid=%-5t  %q\n", gqql.IsValid(q), q)
	}
}

func trunc(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}

func truncIdx(s string, n int) int {
	if len(s) > n {
		return n
	}
	return len(s)
}

File: examples/rust/01_basic_parse/src/main.rs
use qql_core::lexer::Lexer;
use qql_core::parser::Parser;

fn main() {
    let stmt = Parser::parse("CREATE COLLECTION docs HYBRID WITH HNSW (m = 32)").unwrap();
    println!("=== Parsed AST ===");
    println!("{}", &format!("{:#?}", stmt)[..500.min(format!("{:#?}", stmt).len())]);

    println!("\n=== Tokens ===");
    for tok in Lexer::new("QUERY 'vector database' FROM docs LIMIT 10") {
        let tok = tok.unwrap();
        println!("  {:12} {:30}  pos={}", tok.kind.as_str(), tok.text, tok.pos);
    }

    println!("\n=== Validation ===");
    for q in [
        "QUERY 'hello' FROM docs LIMIT 5",
        "CREATE COLLECTION docs",
        "SELECT * FROM docs WHERE id = 1",
        "",
        "BOGUS STUFF",
    ] {
        println!("  valid={:<5}  {:?}", Parser::parse(q).is_ok(), q);
    }
}
